slay/definitions.py

```diff
@@ -0,0 +1,259 @@
+
+BASTEN_APY_SECRET_NAME = "baseten_api_key"
+TRUSS_CONFIG_SLAY_KEY = "slay_metadata"
+
+ENDPOINT_NAME = "run"  # Referring to processor method name exposed as endpoint.
+# Below arg names must correspond to `definitions.ABCProcessor`.
+CONTEXT_ARG_NAME = "context"  # Referring to processors `__init__` signature.
+SELF_ARG_NAME = "self"
+
+GENERATED_CODE_DIR = ".slay_generated"
+PREDICT_ENDPOINT = "/predict"
+PROCESSOR_MODULE = "processor"
+
+
+class APIDefinitonError(TypeError):
+    """Raised when user-defined processors do not adhere to API constraints."""
+
+
+class MissingDependencyError(TypeError):
+    """Raised when a needed resource could not be found or is not defined."""
+
+
+class UsageError(Exception):
+    """Raised when components are not used the expected way at runtime."""
+
+
+class ImageSpec(pydantic.BaseModel):
+    # TODO: this is not stable yet and might change or refer back to truss.
+    base_image: str = "python:3.11-slim"
+    pip_requirements_file: Optional[str] = None
+    pip_requirements: list[str] = []
+    apt_requirements: list[str] = []
+
+
+class Image:
+    """Builder to create image spec."""
+
+    _spec: ImageSpec
+
+    def __init__(self) -> None:
+        self._spec = ImageSpec()
+
+    def pip_requirements_file(self, file_path: str) -> "Image":
+        # TODO: Make this work with relative paths.
+        self._spec.pip_requirements_file = file_path
+        return self
+
+    def pip_requirements(self, requirements: list[str]) -> "Image":
+        self._spec.pip_requirements = requirements
+        return self
+
+    def apt_requirements(self, requirements: list[str]) -> "Image":
+        self._spec.apt_requirements = requirements
+        return self
+
+    def get_spec(self) -> ImageSpec:
+        return self._spec.copy(deep=True)
+
+
+class ComputeSpec(pydantic.BaseModel):
+    # TODO: this is not stable yet and might change or refer back to truss.
+    cpu: str = "1"
+    memory: str = "2Gi"
+    gpu: Optional[str] = None
+
+
+class Compute:
+    """Builder to create compute spec."""
+
+    _spec: ComputeSpec
+
+    def __init__(self) -> None:
+        self._spec = ComputeSpec()
+
+    def cpu(self, cpu: int) -> "Compute":
+        self._spec.cpu = str(cpu)
+        return self
+
+    def memory(self, memory: str) -> "Compute":
+        self._spec.memory = memory
+        return self
+
+    def gpu(self, kind: str, count: int = 1) -> "Compute":
+        self._spec.gpu = f"{kind}:{count}"
+        return self
+
+    def get_spec(self) -> ComputeSpec:
+        return self._spec.copy(deep=True)
+
+
+class AssetSpec(pydantic.BaseModel):
+    # TODO: this is not stable yet and might change or refer back to truss.
+    secrets: dict[str, str] = {}
+    cached: list[Any] = []
+
+
+class Assets:
+    """Builder to create asset spec."""
+
+    _spec: AssetSpec
+
+    def __init__(self) -> None:
+        self._spec = AssetSpec()
+
+    def secret(self, key: str) -> "Assets":
```

Comment:
should this be set_secret? i'm assuming this part is not really fleshed out, but just curious what the long term API is here.

Reply:
Changed to `add_secret`. And yes, this part is not very fleshed out yet and I'll focus on it later.